### PR TITLE
[geo_mean] Fix DP-inconsistent loss aggregation bypassing agg_loss()

### PR DESCRIPTION
**Pull request by dubin555**
## Problem
In the geometric mean loss computation, `torch.mean()` is used directly instead of the `agg_loss()` helper. This bypasses the DP-aware aggregation logic, causing the effective learning rate to scale incorrectly with `dp_size` during multi-GPU training.

## Fix
Replace `torch.mean()` with the proper `agg_loss()` call to ensure consistent loss scaling across data-parallel ranks.

## Testing
- CPU validation confirms the fix uses the correct aggregation path
- No impact on single-GPU training

### CODE REVIEW

File: verl/trainer/ppo/core_algos.py [hunk 1]
```diff
@@ -1987,7 +1988,15 @@ def compute_policy_loss_geo_mean(
...
-    pg_loss = torch.mean(pg_losses)
+    # DP-aware sequence-level aggregation (pg_losses is already per-sequence)
+    seq_mask = (response_mask.sum(dim=-1) > 0).float()
+    global_batch_size = config.global_batch_info.get("global_batch_size", None)
+    dp_size = config.global_batch_info.get("dp_size", 1)
+    if global_batch_size is None:
+        if dp_size > 1:
+            raise ValueError("global_batch_size is required when dp_size > 1 for geo_mean loss")
+        global_batch_size = seq_mask.sum()
+    pg_loss = (pg_losses * seq_mask).sum() / global_batch_size * dp_size
```
[the remark this answers is not in the document]
Comment (dubin555) [reply verl/trainer/ppo/core_algos.py]: Thanks for the review! Updated in 5e0a583 — now delegates to `agg_loss` with `seq-mean-token-sum` mode as suggested, keeping the aggregation logic centralized.

**Comment by gss10282025** [comment]
## [Bug] GMPO geo_mean loss changes when a minibatch is split into more microbatches

### System Info

```text
verl: 24f25b03aa4b54249a273655ebbcce06f484192b
PyTorch: 2.11.0+cu130
Transformers: 5.5.3
Python: 3.12
GPU: NVIDIA GeForce RTX 5090, one visible GPU per process
```

Source: [core_algos.py at 24f25b03](https://github.com/verl-project/verl/blob/24f25b03aa4b54249a273655ebbcce06f484192b/verl/trainer/ppo/core_algos.py).

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [x] My own task or dataset (give details below)

A fixed rollout of eight sequences for one PPO update; no rollout regeneration.

### Reproduction

This reproduces the aggregation problem discussed in #5614 through the FSDP engine itself, with one data-parallel rank.

`compute_policy_loss_geo_mean` ends with `pg_loss = torch.mean(pg_losses)`, a mean over the sequences in the current microbatch, and does not go through `agg_loss` or the global batch metadata. The FSDP loop then calls `backward()` on each microbatch loss with no rescaling, so a minibatch split into eight pieces contributes eight local means instead of one minibatch mean.

Steps:

1. Go through the FSDP engine's `train_batch` → `forward_backward_batch` → `ppo_loss` → `compute_policy_loss_geo_mean` with eight fixed sequences (response length 16, 128 valid response tokens).
2. One SGD update, learning rate `0.02`, gradient clipping `1.0`, KL coefficient `0.001`, same initial model and optimizer state.
3. Compare one microbatch of eight sequences with eight microbatches of one sequence.

| One microbatch vs. eight | Relative L2 difference |
| --- | ---: |
| Pre-clip gradient | `0.8749962` |
| Parameter update | `0.5407175` |

Three repeated runs give the same result, and a second host reproduces the vectors exactly.

### Expected behavior

Each microbatch should contribute its share of the minibatch mean. With one rank:

```text
contribution = local_mean * local_sequence_count / minibatch_sequence_count
```

The denominator has to be the sequence count of the optimizer minibatch. Dividing by the number of microbatches only works when all microbatches have the same number of sequences.

A local weighting like this, applied without changing the eight-microbatch execution, brings the gradient and update differences down to numerical noise and agrees with an independent computation of the objective. The implementation also compensates for the engine's DP gradient averaging, but the numbers above are single-rank. This does not measure downstream reward and does not assume #5614 is part of the tested commit.
